src/llmq/quorums_dkgsessionhandler.cpp

```diff
@@ -0,0 +1,577 @@
+    } else if (strCommand == NetMsgType::QPCOMMITMENT) {
+        pendingPrematureCommitments.PushPendingMessage(pfrom->GetId(), vRecv, MSG_QUORUM_PREMATURE_COMMITMENT);
+    }
+}
+
+bool CDKGSessionHandler::InitNewQuorum(const CBlockIndex* pindexQuorum)
+{
+    curSession = std::make_shared<CDKGSession>(params, evoDb, blsWorker, dkgManager);
+
+    if (!deterministicMNManager->IsDIP3Enforced(pindexQuorum->nHeight) ||
+            !activeMasternodeManager) {
+        return false;
+    }
+
+    auto mns = deterministicMNManager->GetAllQuorumMembers(params.type, pindexQuorum);
+
+    if (!curSession->Init(pindexQuorum, mns, activeMasternodeManager->GetProTx())) {
+        LogPrintf("CDKGSessionHandler::%s -- quorum initialiation failed for %s\n", __func__, curSession->params.name);
+        return false;
+    }
+
+    return true;
+}
+
+std::pair<QuorumPhase, uint256> CDKGSessionHandler::GetPhaseAndQuorumHash() const
+{
+    LOCK(cs);
+    return std::make_pair(phase, quorumHash);
+}
+
+class AbortPhaseException : public std::exception {
+};
+
+void CDKGSessionHandler::WaitForNextPhase(QuorumPhase curPhase,
+                                          QuorumPhase nextPhase,
+                                          const uint256& expectedQuorumHash,
+                                          const WhileWaitFunc& runWhileWaiting)
+{
+    LogPrint(BCLog::DKG, "CDKGSessionHandler::%s -- %s - starting, curPhase=%d, nextPhase=%d\n", __func__, params.name, curPhase, nextPhase);
+
+    while (true) {
+        if (stopRequested || ShutdownRequested()) {
+            throw AbortPhaseException();
+        }
+        auto p = GetPhaseAndQuorumHash();
+        if (!expectedQuorumHash.IsNull() && p.second != expectedQuorumHash) {
+            throw AbortPhaseException();
+        }
+        if (p.first == nextPhase) {
+            break;
+        }
+        if (curPhase != QuorumPhase_None && p.first != curPhase) {
+            LogPrint(BCLog::DKG, "CDKGSessionHandler::%s -- %s - aborting due unexpected phase change\n", __func__, params.name);
+            throw AbortPhaseException();
+        }
+        if (!runWhileWaiting()) {
+            MilliSleep(100);
+        }
+    }
+
+    // !TODO: reset/update local session status
+
+    LogPrint(BCLog::DKG, "CDKGSessionHandler::%s -- %s - done, curPhase=%d, nextPhase=%d\n", __func__, params.name, curPhase, nextPhase);
+
+}
+
+void CDKGSessionHandler::WaitForNewQuorum(const uint256& oldQuorumHash)
+{
+    LogPrint(BCLog::DKG, "CDKGSessionHandler::%s -- %s - starting\n", __func__, params.name);
+
+    while (true) {
+        if (stopRequested || ShutdownRequested()) {
+            throw AbortPhaseException();
+        }
+        auto p = GetPhaseAndQuorumHash();
+        if (p.second != oldQuorumHash) {
+            return;
+        }
+        MilliSleep(100);
+    }
+
+    LogPrint(BCLog::DKG, "CDKGSessionHandler::%s -- %s - done\n", __func__, params.name);
+}
```

Comment:
the `return;` at line 222 should be a `break;`. Otherwise the `LogPrint` is unreachable code.

Reply:
nice catch